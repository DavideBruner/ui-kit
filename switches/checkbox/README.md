# Switches: Checkbox

## Usage

```js
import { Checkbox } from '@commercetools-frontend/ui-kit/switches';
```

#### Description

A `Checkbox` is an element which indicates an checked or non checked status and
notifies with a value accordingly.

#### Usage

```js
import { Checkbox } from '@commercetools-frontend/ui-kit/switches';

<Checkbox value="foo-radio-value" onChange={..}>
  What ever option
</Checkbox>
<Checkbox value="bar-radio-value" isDisabled={true} onChange={..}>
  Another option
</Checkbox>
```

#### Properties

| Props             | Type     | Required | Values | Default | Description                                                                                                                                                                                                                                                                             |
| ----------------- | -------- | :------: | ------ | ------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `children`        | `string` |    -     | -      | -       | The descriptive text of the Checkbox                                                                                                                                                                                                                                                    |
| `isDisabled`      | `bool`   |    -     | -      | `false` | Disables the Checkbox                                                                                                                                                                                                                                                                   |
| `isIndeterminate` | `bool`   |    -     | -      | `false` | The indeterminate state is visual only. Its used for nested checkboxes. Each checkbox may have child checkboxes. If all those children are checked, it may be checked. If none are checked, it is unchecked. If some of them are checked, then `isIndeterminate` state should be `true` |
| `isChecked`       | `bool`   |    -     | -      | `false` | Checks the Checkbox                                                                                                                                                                                                                                                                     |
| `value`           | `string` |    -     | -      | -       | The value of the option passed to `onChange` if not checked                                                                                                                                                                                                                             |
| `hasError`        | `bool`   |    -     | -      | -       | Indicates that the checkbox has an error                                                                                                                                                                                                                                                |
| `onChange`        | `func`   |    âœ…    | -      | -       | What will trigger whenever an `Radio.Option` is clicked. Trigged with `undefined` is checked otherwise the value                                                                                                                                                                        |
