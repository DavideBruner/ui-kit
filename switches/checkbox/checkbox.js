import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import filterDataAttributes from '../../utils/filter-data-attributes';
import Text from '../../typography/text';
import Spacings from '../../materials/spacings';
import withMouseOverState from '../../hocs/with-mouse-over-state';
import Icons from './icons';
import styles from './checkbox.mod.css';

const getTextTone = ({ isDisabled, hasError }) => {
  if (isDisabled) return 'secondary';
  if (hasError) return 'negative';
  return undefined;
};
export class Checkbox extends React.PureComponent {
  static displayName = 'Checkbox';
  static propTypes = {
    name: PropTypes.string,
    value: PropTypes.string,
    isChecked: PropTypes.bool,
    isIndeterminate: PropTypes.bool,
    onChange: PropTypes.func.isRequired,
    isDisabled: PropTypes.bool,
    hasError: PropTypes.bool,
    children: PropTypes.node,
    // HoC
    isMouseOver: PropTypes.bool.isRequired,
    handleMouseOver: PropTypes.func.isRequired,
    handleMouseOut: PropTypes.func.isRequired,
  };

  static defaultProps = {
    isChecked: false,
    isDisabled: false,
    hasError: false,
  };

  render() {
    return (
      <div
        onMouseOver={this.props.handleMouseOver}
        onMouseOut={this.props.handleMouseOut}
        {...filterDataAttributes(this.props)}
      >
        <label
          className={classnames(styles.labelWrapper, {
            [styles.labelWrapperDisabled]: this.props.isDisabled,
            [styles.labelWrapperError]: this.props.hasError,
          })}
        >
          <Spacings.Inline alignItems="center">
            <div
              className={classnames(styles.checkboxWrapper, {
                [styles.isDisabled]: this.props.isDisabled,
                [styles.hasError]: this.props.hasError,
              })}
            >
              {(() => {
                if (this.props.isIndeterminate) return <Icons.Indeterminate />;
                else if (this.props.isChecked) return <Icons.Checked />;
                return <Icons.Unchecked />;
              })()}
            </div>
            {this.props.children && (
              <Text.Body
                tone={getTextTone({
                  isDisabled: this.props.isDisabled,
                  hasError: this.props.hasError,
                })}
              >
                {this.props.children}
              </Text.Body>
            )}
            <input
              className={styles.inputWrapper}
              name={this.props.name}
              value={this.props.value}
              onChange={this.props.onChange}
              disabled={this.props.isDisabled}
              checked={this.props.isChecked && !this.props.isIndeterminate}
              type="checkbox"
            />
          </Spacings.Inline>
        </label>
      </div>
    );
  }
}

export default withMouseOverState(Checkbox);
