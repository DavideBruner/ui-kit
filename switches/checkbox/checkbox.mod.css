@import '../../materials/spacings.mod.css';
@import '../../materials/colors.mod.css';

.inputWrapper {
  display: none;
}

.labelWrapper {
  cursor: pointer;
}

.labelWrapperDisabled {
  cursor: not-allowed;
}

/* Text  */

/* isDisabled */

.textWrapperDisabled {
  color: var(--color-primaryGrey);
}

/* hasError */

.textWrapperError {
  color: var(--color-red);
}

/* Checkbox */

/* isDisabled */

.isDisabled svg g[id='borderAndContent'] > :not([id='border']) {
  fill: var(--color-primaryGrey);
}

.isDisabled svg [id='borderAndContent'] > [id='border'] {
  stroke: var(--color-primaryGrey);
}

/* hasError */

.hasError svg g[id='borderAndContent'] > :not([id='border']) {
  fill: var(--color-red);
}

.hasError svg [id='borderAndContent'] > [id='border'] {
  stroke: var(--color-red);
}

/* Hover */

.labelWrapper:not(.labelWrapperDisabled, .labelWrapperError):hover
  svg
  [id='borderAndContent'] > [id='border'] {
  stroke: var(--color-primaryGreen);
}
