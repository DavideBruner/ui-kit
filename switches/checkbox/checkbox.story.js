import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, boolean, text } from '@storybook/addon-knobs';
import withReadme from 'storybook-readme/with-readme';
import Spacings from '../../materials/spacings';
import Section from '../../.storybook/decorators/section';
import Readme from './README.md';
import Checkbox from './';

storiesOf('Switches', module)
  .addDecorator(withKnobs)
  .addDecorator(withReadme(Readme))
  .add('Checkbox', () => (
    <Section>
      <Spacings.Stack>
        <Checkbox
          isDisabled={boolean('isFirstCheckboxDisabled', false)}
          isChecked={boolean('isFirstCheckboxChecked', true)}
          onChange={action('onChange: first')}
          value="foo-value"
        >
          {text('Label of primary action', 'Primary checkbox')}
        </Checkbox>
        <Checkbox
          isDisabled={boolean('isSecondCheckboxDisabled', false)}
          onChange={action('onChange: second')}
          value="bar-value"
        >
          {text('Label of second action', 'Second checkbox')}
        </Checkbox>
        <Checkbox
          isDisabled={boolean('isThirdCheckboxDisabled', true)}
          onChange={action('onChange: third')}
          value="baz-value"
        >
          {text('Label of third action', 'Third checkbox')}
        </Checkbox>
        <Checkbox
          isDisabled={boolean('isFourthCheckboxDisabled', false)}
          isChecked={boolean('isFourthCheckboxChecked', false)}
          onChange={action('onChange: fourth')}
          value="x-value"
          isIndeterminate={boolean('isFourthCheckboxIndeterminate', true)}
        >
          {text('Label of fourth action', 'Fourth checkbox')}
        </Checkbox>
      </Spacings.Stack>
    </Section>
  ));
