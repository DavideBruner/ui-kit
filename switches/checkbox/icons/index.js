// NOTE: explicit names for component names
import CheckboxDefaultIcon from './checkbox-default.svg';
import CheckboxDefaultHoverIcon from './checkbox-hover.svg';
import CheckboxDefaultErrorIcon from './checkbox-error.svg';
import CheckboxCheckedIcon from './checkbox-checked.svg';
import CheckboxCheckedHoverIcon from './checkbox-checked-hover.svg';
import CheckboxCheckedErrorIcon from './checkbox-checked-error.svg';
import CheckboxDisabledIcon from './checkbox-disabled.svg';
import CheckboxCheckedDisabledIcon from './checkbox-checked-disabled.svg';
import CheckboxIndeterminateIcon from './checkbox-indeterminate.svg';
import CheckboxIndeterminateHoverIcon from './checkbox-indeterminate-hover.svg';
import CheckboxIndeterminateDisabledIcon from './checkbox-indeterminate-disabled.svg';
import CheckboxIndeterminateErrorIcon from './checkbox-indeterminate-error.svg';

export default {
  Default: CheckboxDefaultIcon,
  Hover: CheckboxDefaultHoverIcon,
  Error: CheckboxDefaultErrorIcon,
  Checked: CheckboxCheckedIcon,
  CheckedHover: CheckboxCheckedHoverIcon,
  CheckedError: CheckboxCheckedErrorIcon,
  Disabled: CheckboxDisabledIcon,
  CheckedDisabled: CheckboxCheckedDisabledIcon,
  Indeterminate: CheckboxIndeterminateIcon,
  IndeterminateHover: CheckboxIndeterminateHoverIcon,
  IndeterminateDisabled: CheckboxIndeterminateDisabledIcon,
  IndeterminateError: CheckboxIndeterminateErrorIcon,
};
