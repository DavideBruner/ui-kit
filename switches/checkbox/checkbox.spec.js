import React from 'react';
import { shallow } from 'enzyme';
import { Checkbox } from './checkbox';

jest.mock('invariant');

const createTestProps = custom => ({
  name: 'bar',
  value: 'foo',
  isDisabled: false,
  isChecked: false,
  onChange: jest.fn(),

  // HoC
  handleMouseOver: jest.fn(),
  handleMouseOut: jest.fn(),
  isMouseOver: false,
  ...custom,
});

describe('<Checkbox>', () => {
  describe('rendering', () => {
    let props;
    let wrapper;
    describe('without children', () => {
      beforeEach(() => {
        props = createTestProps();

        wrapper = shallow(<Checkbox {...props} />);
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should supply `onChange` to the `input`', () => {
        expect(wrapper.find('input')).toHaveProp('onChange', props.onChange);
      });

      describe('with `data-*`', () => {
        beforeEach(() => {
          props = createTestProps({
            'data-test': 'foo-attribute-select-me-from-e2e',
          });
          wrapper = shallow(<Checkbox {...props} />);
        });
        it('should match snapshot', () => {
          expect(wrapper).toMatchSnapshot();
        });
      });

      describe('when disabled', () => {
        beforeEach(() => {
          props = createTestProps({ isDisabled: true });

          wrapper = shallow(<Checkbox {...props} />);
        });

        it('should disable the `input`', () => {
          expect(wrapper.find('input')).toHaveProp(
            'disabled',
            props.isDisabled
          );
        });
      });

      describe('when intermediate', () => {
        beforeEach(() => {
          props = createTestProps({ isIndeterminate: true });

          wrapper = shallow(
            <div>
              <Checkbox {...props} />
            </div>
          );
        });

        it('should contain `isIndeterminate` prop', () => {
          expect(wrapper.find(Checkbox)).toHaveProp(
            'isIndeterminate',
            props.isIndeterminate
          );
        });
      });

      describe('with error', () => {
        beforeEach(() => {
          props = createTestProps({ hasError: true });

          wrapper = shallow(
            <div>
              <Checkbox {...props} />
            </div>
          );
        });

        it('should prop `hasError` be true', () => {
          expect(wrapper.find(Checkbox)).toHaveProp('hasError', true);
        });
      });

      describe('when checked', () => {
        beforeEach(() => {
          props = createTestProps({ isDisabled: true });

          wrapper = shallow(<Checkbox {...props} />);
        });

        it('should check the `input`', () => {
          expect(wrapper.find('input')).toHaveProp('checked', props.isChecked);
        });
      });
    });

    describe('with children', () => {
      beforeEach(() => {
        props = createTestProps();

        wrapper = shallow(<Checkbox {...props}>{'Some label'}</Checkbox>);
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});

describe('callbacks', () => {
  const createEvent = ({ isChecked = true } = {}) => ({
    target: {
      checked: isChecked,
    },
  });

  describe('when changing', () => {
    let props;
    describe('when checked', () => {
      beforeEach(() => {
        props = createTestProps({ isChecked: true });

        const wrapper = shallow(<Checkbox {...props} />);
        wrapper
          .find('input')
          .simulate('change', createEvent({ isChecked: props.isChecked }));
      });

      it('should invoke `onChange`', () => {
        expect(props.onChange).toHaveBeenCalled();
      });

      it('should invoke `onChange` with `checked`', () => {
        expect(props.onChange).toHaveBeenCalledWith(
          expect.objectContaining({
            target: expect.objectContaining({
              checked: props.isChecked,
            }),
          })
        );
      });

      it('should invoke `onChange` with `event`', () => {
        expect(props.onChange).toHaveBeenCalledWith(
          expect.objectContaining({
            target: expect.any(Object),
          })
        );
      });
    });

    describe('when not checked', () => {
      beforeEach(() => {
        props = createTestProps({ isChecked: false });

        const wrapper = shallow(<Checkbox {...props} />);

        wrapper
          .find('input')
          .simulate('change', createEvent({ isChecked: props.isChecked }));
      });

      it('should invoke `onChange`', () => {
        expect(props.onChange).toHaveBeenCalled();
      });

      it('should invoke `onChange` with `checked`', () => {
        expect(props.onChange).toHaveBeenCalledWith(
          expect.objectContaining({
            target: expect.objectContaining({
              checked: props.isChecked,
            }),
          })
        );
      });

      it('should invoke `onChange` with `event`', () => {
        expect(props.onChange).toHaveBeenCalledWith(
          expect.objectContaining({
            target: expect.any(Object),
          })
        );
      });
    });
  });
});
