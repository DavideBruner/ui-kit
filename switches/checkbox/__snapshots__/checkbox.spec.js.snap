// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox> rendering with children should match snapshot 1`] = `
<div
  onMouseOut={[MockFunction]}
  onMouseOver={[MockFunction]}
>
  <label
    className="labelWrapper"
  >
    <Inline
      alignItems="center"
      scale="s"
    >
      <test-file-stub />
      <span
        className=""
      >
        Some label
      </span>
      <input
        checked={false}
        className="inputWrapper"
        disabled={false}
        name="bar"
        onChange={[MockFunction]}
        type="checkbox"
        value="foo"
      />
    </Inline>
  </label>
</div>
`;

exports[`<Checkbox> rendering without children should match snapshot 1`] = `
<div
  onMouseOut={[MockFunction]}
  onMouseOver={[MockFunction]}
>
  <label
    className="labelWrapper"
  >
    <Inline
      alignItems="center"
      scale="s"
    >
      <test-file-stub />
      <input
        checked={false}
        className="inputWrapper"
        disabled={false}
        name="bar"
        onChange={[MockFunction]}
        type="checkbox"
        value="foo"
      />
    </Inline>
  </label>
</div>
`;

exports[`<Checkbox> rendering without children with \`data-*\` should match snapshot 1`] = `
<div
  data-test="foo-attribute-select-me-from-e2e"
  onMouseOut={[MockFunction]}
  onMouseOver={[MockFunction]}
>
  <label
    className="labelWrapper"
  >
    <Inline
      alignItems="center"
      scale="s"
    >
      <test-file-stub />
      <input
        checked={false}
        className="inputWrapper"
        disabled={false}
        name="bar"
        onChange={[MockFunction]}
        type="checkbox"
        value="foo"
      />
    </Inline>
  </label>
</div>
`;
